optimizely/odp/lru_cache.py

```diff
@@ -0,0 +1,133 @@
+# limitations under the License.
+
+
+from __future__ import annotations
+from dataclasses import dataclass, field
+import threading
+from time import time
+from collections import OrderedDict
+from typing import Optional, Generic, TypeVar, Hashable
+from sys import version_info
+
+if version_info < (3, 8):
+    from typing_extensions import Protocol
+else:
+    from typing import Protocol  # type: ignore
+
+
+# generic type definitions for LRUCache parameters
+K = TypeVar('K', bound=Hashable, contravariant=True)
+V = TypeVar('V')
+
+
+class LRUCache(Generic[K, V]):
+    """Least Recently Used cache that invalidates entries older than the timeout."""
+
+    def __init__(self, capacity: int, timeout_in_secs: float):
+        self.lock = threading.Lock()
+        self.map: OrderedDict[K, CacheElement[V]] = OrderedDict()
+        self.capacity = capacity
+        self.timeout = timeout_in_secs
+
+    def lookup(self, key: K) -> Optional[V]:
+        """Return the non-stale value associated with the provided key and move the
+        element to the end of the cache. If the selected value is stale, remove it from
+        the cache and clear the entire cache if stale.
+        """
+        if self.capacity <= 0:
+            return None
+
+        with self.lock:
+            if key not in self.map:
+                return None
+
+            self.map.move_to_end(key)
+            element = self.map[key]
+
+            if element._is_stale(self.timeout):
+                del self.map[key]
+                if self._all_stale():
+                    self.map.clear()
+                return None
+
+        return element.value
+
+    def save(self, key: K, value: V) -> None:
+        """Insert and/or move the provided key/value pair to the most recent end of the cache.
+        If the cache grows beyond the cache capacity, the least recently used element will be
+        removed.
+        """
+        if self.capacity <= 0:
+            return
+
+        with self.lock:
+            if key in self.map:
+                self.map.move_to_end(key)
+
+            self.map[key] = CacheElement(value)
+
+            if len(self.map) > self.capacity:
+                self.map.popitem(last=False)
+
+    def _all_stale(self) -> bool:
+        """Returns True if the timeout has passed since the most recent element's timestamp."""
+        newest_element = next(reversed(self.map.values()))
+        return newest_element._is_stale(self.timeout)
+
+    def reset(self) -> None:
+        """ Clear the cache."""
+        if self.capacity <= 0:
+            return
+        with self.lock:
+            self.map.clear()
+
+    def peek(self, key: K) -> Optional[V]:
+        """Returns the value associated with the provided key without updating the cache."""
+        if self.capacity <= 0:
+            return None
+        with self.lock:
+            element = self.map.get(key)
+        return element.value if element is not None else element
```

Comment:
```suggestion
        return element.value if element is not None else None
```
Not sure for python, but it looks better in this way for typing.